File: integrations/langchain-python/xache_langchain/retriever.py
"""
Xache Retriever for LangChain
Semantic memory retrieval for RAG pipelines
"""

import os
from typing import List, Optional, Dict, Any
from langchain.schema import BaseRetriever, Document
from langchain.callbacks.manager import CallbackManagerForRetrieverRun
from pydantic import Field

from xache import XacheClient
from ._async_utils import run_sync


class XacheRetriever(BaseRetriever):
    """
    LangChain retriever backed by Xache Protocol.

    Retrieves semantically relevant memories for RAG pipelines.
    Each retrieval is paid via x402 micropayments.

    Example:
        ```python
        from xache_langchain import XacheRetriever
        from langchain.chains import RetrievalQA

        retriever = XacheRetriever(
            wallet_address="0x...",
            private_key="0x...",
            k=5
        )

        qa = RetrievalQA.from_chain_type(
            llm=llm,
            retriever=retriever
        )
        ```

    For collective intelligence retrieval:
        ```python
        retriever = XacheRetriever(
            wallet_address="0x...",
            private_key="0x...",
            collective_id="research-insights",  # Query collective
            k=10
        )
        ```
    """

    # Xache configuration
    api_url: str = Field(
        default_factory=lambda: os.environ.get("XACHE_API_URL", "https://api.xache.xyz")
    )
    wallet_address: str = Field(...)
    private_key: str = Field(...)
    chain: str = Field(default="base")

    # Retrieval configuration
    k: int = Field(default=5, description="Number of documents to retrieve")
    collective_id: Optional[str] = Field(
        default=None,
        description="Collective ID for collective intelligence queries"
    )
    include_metadata: bool = Field(default=True)
    min_relevance: float = Field(
        default=0.0,
        description="Minimum relevance score (0-1)"
    )

    # Internal
    _client: Optional[XacheClient] = None

    class Config:
        arbitrary_types_allowed = True
        underscore_attrs_are_private = True

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self._init_client()

    def _init_client(self):
        """Initialize Xache client"""
        chain_prefix = "sol" if self.chain == "solana" else "evm"
        did = f"did:agent:{chain_prefix}:{self.wallet_address.lower()}"

        self._client = XacheClient(
            api_url=self.api_url,
            did=did,
            private_key=self.private_key,
        )

    def _get_relevant_documents(
        self,
        query: str,
        *,
        run_manager: CallbackManagerForRetrieverRun,
    ) -> List[Document]:
        """Retrieve relevant documents from Xache"""

        async def _retrieve():
            async with self._client as client:
                if self.collective_id:
                    # Query collective intelligence
                    result = await client.collective.query(
                        collective_id=self.collective_id,
                        query=query,
                        limit=self.k
                    )

                    documents = []
                    for item in result.get("results", []):
                        content = item.get("content", "")
                        metadata = {
                            "source": "xache_collective",
                            "collective_id": self.collective_id,
                            "contributor_did": item.get("contributor_did"),
                            "relevance": item.get("relevance", 0),
                            "receipt_id": item.get("receipt_id"),
                        }
                        if self.include_metadata:
                            metadata.update(item.get("metadata", {}))

                        if item.get("relevance", 1) >= self.min_relevance:
                            documents.append(Document(
                                page_content=content,
                                metadata=metadata
                            ))

                    return documents
                else:
                    # Query personal memories
                    result = await client.memory.retrieve(
                        query=query,
                        limit=self.k
                    )

                    documents = []
                    # Fix: Correctly normalize result - API returns {"memories": [...]}
                    memories = result.get("memories", []) if isinstance(result, dict) else []

                    for mem in memories:
                        content = mem.get("content", "")
                        metadata = {
                            "source": "xache_memory",
                            "storage_key": mem.get("storage_key"),
                            "created_at": mem.get("created_at"),
                            "receipt_id": mem.get("receipt_id"),
                        }
                        if self.include_metadata:
                            metadata.update(mem.get("metadata", {}))

                        documents.append(Document(
                            page_content=content,
                            metadata=metadata
                        ))

                    return documents

        return run_sync(_retrieve())

    async def _aget_relevant_documents(
        self,
        query: str,
        *,
        run_manager: CallbackManagerForRetrieverRun,
    ) -> List[Document]:
        """Async retrieve relevant documents"""
        async with self._client as client:
            if self.collective_id:
                result = await client.collective.query(
                    collective_id=self.collective_id,
                    query=query,
                    limit=self.k
                )

                documents = []
                for item in result.get("results", []):
                    content = item.get("content", "")
                    metadata = {
                        "source": "xache_collective",
                        "collective_id": self.collective_id,
                        "contributor_did": item.get("contributor_did"),
                        "relevance": item.get("relevance", 0),
                    }
                    if self.include_metadata:
                        metadata.update(item.get("metadata", {}))

                    if item.get("relevance", 1) >= self.min_relevance:
                        documents.append(Document(
                            page_content=content,
                            metadata=metadata
                        ))

                return documents
            else:
                result = await client.memory.retrieve(
                    query=query,
                    limit=self.k
                )

                documents = []
                # Fix: Correctly normalize result - API returns {"memories": [...]}
                memories = result.get("memories", []) if isinstance(result, dict) else []

                for mem in memories:
                    content = mem.get("content", "")
                    metadata = {
                        "source": "xache_memory",
                        "storage_key": mem.get("storage_key"),
                        "created_at": mem.get("created_at"),
                    }
                    if self.include_metadata:
                        metadata.update(mem.get("metadata", {}))

                    documents.append(Document(
                        page_content=content,
                        metadata=metadata
                    ))

                return documents

File: integrations/langchain-ts/src/retriever.ts
/**
 * Xache Retriever for LangChain.js
 * Memory retrieval for RAG pipelines with verifiable receipts
 */

import { BaseRetriever, BaseRetrieverInput } from '@langchain/core/retrievers';
import { Document } from '@langchain/core/documents';
import { CallbackManagerForRetrieverRun } from '@langchain/core/callbacks/manager';
import { XacheClient, DID, type XacheSigner, type XacheWalletProvider } from '@xache/sdk';

export interface XacheRetrieverConfig extends BaseRetrieverInput {
  /** Wallet address for authentication */
  walletAddress: string;
  /** Private key for signing (optional if signer or walletProvider is provided) */
  privateKey?: string;
  /** External signer (alternative to privateKey) */
  signer?: XacheSigner;
  /** Wallet provider for lazy signer resolution */
  walletProvider?: XacheWalletProvider;
  /** Encryption key for use with external signers */
  encryptionKey?: string;
  /** Number of documents to retrieve */
  k?: number;
  /** Filter by context */
  context?: string;
  /** API URL (defaults to https://api.xache.xyz) */
  apiUrl?: string;
  /** Chain: 'base' or 'solana' */
  chain?: 'base' | 'solana';
}

/**
 * Retriever that fetches documents from Xache memory storage.
 *
 * Use this for RAG pipelines with persistent, verifiable document storage.
 *
 * @example
 * ```typescript
 * import { XacheRetriever } from '@xache/langchain';
 * import { ChatOpenAI } from '@langchain/openai';
 * import { RetrievalQAChain } from 'langchain/chains';
 *
 * const retriever = new XacheRetriever({
 *   walletAddress: '0x...',
 *   privateKey: '0x...',
 *   k: 5,
 * });
 *
 * const qa = RetrievalQAChain.fromLLM(new ChatOpenAI(), retriever);
 * const result = await qa.call({ query: 'What do you know about X?' });
 * ```
 */
export class XacheRetriever extends BaseRetriever {
  lc_namespace = ['xache', 'retriever'];

  static lc_name() {
    return 'XacheRetriever';
  }

  private client: XacheClient;
  private k: number;
  private filterContext?: string;

  constructor(config: XacheRetrieverConfig) {
    super(config);

    const chainPrefix = config.chain === 'solana' ? 'sol' : 'evm';
    const did = `did:agent:${chainPrefix}:${config.walletAddress.toLowerCase()}` as DID;

    this.client = new XacheClient({
      apiUrl: config.apiUrl || 'https://api.xache.xyz',
      did,
      privateKey: config.privateKey,
      signer: config.signer,
      walletProvider: config.walletProvider,
      encryptionKey: config.encryptionKey,
    });

    this.k = config.k ?? 5;
    this.filterContext = config.context;
  }

  async _getRelevantDocuments(
    _query: string,
    _runManager?: CallbackManagerForRetrieverRun
  ): Promise<Document[]> {
    // Use list method to get memories filtered by context
    const result = await this.client.memory.list({
      context: this.filterContext,
      limit: this.k,
    });

    const memories = result.memories || [];

    return memories.map(
      (m) =>
        new Document({
          pageContent: m.context || '',
          metadata: {
            storageKey: m.storage_key,
            context: m.context,
            tier: m.storage_tier,
            createdAt: m.created_at,
            size: m.size_bytes,
          },
        })
    );
  }
}
